### PR TITLE
feat(merge-loader): introduced ConfigMergedLoader

### PR DESCRIPTION
**Pull request by zjx20**
The idea was borrowed from the [config](https://github.com/railsconfig/config) ruby gem.

We usually have different versions of config for different environments or for different people. With `ConfigMergedLoader`, everyone can setup a localized config (a.k.a. `config.local.json`) for their own use. Further more, `ConfigMergedLoader` is smart enough to ignore missed config files (but not all of them), so that we don't need to commit `config.local.json` to the VCS.

### CODE REVIEW

**Comment by fulls1z3** [comment]
Hi @zjx20, this is an awesome contribution to the project. Actually this idea was in my mind already, simply couldn't figure it out when to implement.

I'll review your PR and going to merge it with @nglibs/config asap.

Meanwhile, feel free to have a look at the other utilities under @nglibs scope:

- [@nglibs/meta](https://github.com/nglibs/meta)
- [@nglibs/i18n-router](https://github.com/nglibs/i18n-router)

**Comment by zjx20** [comment]
Sure, I was looking for a config library for my project, and just found @nglibs/config :)

Btw, I have no idea why the `typescript-object-utils` package is introducing those errors. Maybe we have to maintain our own version of `mergeDeep()`.

**Comment by zjx20** [comment]
The best alternate of `mergeDeep()` I could find was https://lodash.com/docs#mergeWith . Also, since the use case is configuration, I prefer not to treat arrays as objects. That said:
```
let a = {'a': [1,2,3]};
let b = {'a': [4,5]};
let c = merge(a, b);
console.log(c);

// should output:
//   {'a': [4,5]}
//
// but NOT:
//   {'a': [4,5,3]}
//   {'a': [1,2,3,4,5]}
```

@fulls1z3 what do you think?

**Comment by fulls1z3** [comment]
I think you aim is to merge values under the same keys.

I gave it a try on JS Bin, with the following input:

```
function customizer(objValue, srcValue) {
  if (_.isArray(objValue)) {
    return objValue.concat(srcValue);
  }
}

var a = {'a': [1,2,3]};
var b = {'a': [4,5]};
var c = _.mergeWith(a, b, customizer);

console.log(c);
```

And it resulted with:

```
[object Object] {
  a: [1, 2, 3, 4, 5]
}
```

Note: You'll need to give a customizer function in order to tell `mergeWith` how to merge.

**Comment by fulls1z3** [comment]
And if we talk about `typescript-object-utils`, I suggest you to better stick on `lodash`. If you check the fellow projects above, you'll notice their usage (simplicity, efficiency, and tests-friendliness) as well.

**Comment by zjx20** [comment]
Good to know (yeah, I'm quite new to the modern js world). I will make changes to use `lodash` instead.

And, actually I would like to treat arrays as value (e.g. numbers or strings), just simply overwrite the value if same key was meet. My proposal is:
```
function customizer(objValue, srcValue) {
  if (_.isArray(objValue)) {
    return srcValue;
  }
}

var a = {
  'num': 123,
  'str': 'str1',
  'arr': [1,2,3],
  'obj': {'v1': 1, 'v2': 2},
  'num1': 999,
};
var b = {
  'num': 456,
  'str': 'str2',
  'arr': [4,5],
  'obj': {'v2': 123, 'v3': 3},
  'new_key': 'added',
};
var c = _.mergeWith(a, b, customizer);

console.log(c);
```

Output should be:
```
{
  // same key, number values should be overwritten with the value from `b`
  'num': 456,

  // same key, string values should be overwritten with the value from `b`
  'str': 'str2',

  // same key, array values should be overwritten with the value from `b`
  'arr': [4,5],

  // same key, object values should be merged recursively
  'obj': {'v1': 1, 'v2': 123, 'v3': 3},

  // for keys don't exist in `b`, simply do nothing
  'num1': 999,

  // for keys don't exist in `a`,  simply add them to `a`
  'new_key': 'added',
}
```

Are you ok with that?

**Comment by fulls1z3** [comment]
Hi @zjx20, I reviewed your suggestion.

I see you'd like to merge config values from different sources under the same key.

This approach would work only in one case: when all the value under a certain key is an array. In this case, the types of the values must also be the same.

However, generating properties below the keys' scope (such as `num`, `str`, `arr`, `obj`) wouldn't be a consistent approach.

Also when a value is supplied with the same key, it should have been overwritten by the new config file/object.

For instance, let's say @nglibs/config processes `source1.json` and then `source2.json`

#### source1.json
```
{
  "availablePorts": [3000],
  "currentPort": 3000,
  "availableLanguages": ['en'],
  "defaultLanguage": 'en'
}
```

#### source2.json
```
{
  "availablePorts": [4200, 3600],
  "currentPort": 3600,
  "availableLanguages": ['fr', 'de']
}
```

#### result
```
{
  "availablePorts": [3000, 4200, 3600],
  "currentPort": 3600,
  "availableLanguages": ['en', 'fr', 'de'],
  "defaultLanguage": 'en'
}
```

Hope it clarifies how the config service should work.

**Comment by zjx20** [comment]
I agree that concatenating arrays is a potential use case, but it's still possible that user want the value (no matter it's an `array` or an `object`) to be entirely overwritten. I can't show an example for that, but I just thought it's possible.

The semantics of merging arrays is ambiguous. I think that is the root of the problem. We can use some tricks to make the semantics to be more precise. For example, a key ends with `+` (e.g. `{"availablePorts+": [4200, 3600]}`) means the value should be concatenated together; a key ends with `=` means it should overwrite the source. The brilliant idea is from [gyp](https://gyp.gsrc.io/docs/InputFormatReference.md#Merge-Basics). It's somehow complicated and shouldn't be implemented in this PR.

However, I still think the default policy of merging array values should be `overwriting`. Because it's more expressive than `concatenating`.

Let's say, the default policy of merging arrays is `overwriting`, now we have a default config file:
```
// config.json
{
  "availablePorts": [3000],
  "currentPort": 3000,
  "availableLanguages": ['en'],
  "defaultLanguage": 'en'
}
```

and then we want the final config object to be:
```
// the final config object
{
  "availablePorts": [3000, 4200, 3600],  // append two numbers to the field
  "currentPort": 3600,
  "availableLanguages": ['fr', 'de'],  // remove 'en' and then add 'fr' and 'de'
  "defaultLanguage": 'fr'
}
```

we can write the localized config file as:
```
// config.local.json
// just rewriting all fields to achieve the goal
{
  "availablePorts": [3000, 4200, 3600],
  "currentPort": 3600,
  "availableLanguages": ['fr', 'de'],
  "defaultLanguage": 'fr'
}
```

But if `concatenating` is the default policy, we can't write such a `config.local.json` to make things work as expected.

**Comment by fulls1z3** [comment]
Two minds are always better than one, and now things are more or less clear. Let's overwrite values, not merge. This way, the `config.local.json` will have a meaning ;)

I'll keep an eye about the idea from **gyp** in future releases.

**Comment by unlight** [comment]
Hi guys, I consider `nglibs/config` for my project. But I totally disagree how ConfigMergedLoader is implemented. Mostly I am against lodash, it brings additional 540 kB. Even lodash.merge gives 58.8 kB overhead. This is not pragmatic.
Consider `Object.assign` or own tiny merge function (if recursive merge needed)

Also I need additional methods to config service, so anyway, I'm going to fork this 🐤

**Comment by fulls1z3** [comment]
Hi @unlight, first of all thanks for your comments as well as the fork. I had a look on it, and it seems that `@nglibs/config` can benefit from your contributions. Please feel free to send your PR's too, that would be awesome.

If we talk about **lodash**, it really provides code cleanliness together with great functionality (*array operations, manipulating values, building composite funcs, etc*) for a price: increased bundle size.

In general, I am against such an overhead (*inflating the bundle size with libs*) too. However, I think one or more of techniques below would keep the difference on bundle size almost at a **negligible level**:
- excluding **lodash** from `@nglibs/config` bundle, and mark it as a `peerDependency`, so the app consuming the `@nglibs/config` can manage how to import **lodash**.
- using **lodash.merge**, or better importing required modules (in our case, `merge`) from **lodash**, or **really better** simply importing **lodash** and let webpack apply **uglify**/**tree-shaking** to the bundle.
- using [lodash-webpack-plugin](https://github.com/lodash/lodash-webpack-plugin)
- using **compression** (*gzip*) at the server
- if possible, importing **lodash** and/or similar libs from their **CDN**, and let the **browser** load them from **cache**

**Comment by zjx20** [comment]
I don't have a strong opinion on lodash, but I don't intent to tweak this PR any further since I don't know how to do.

@fulls1z3 By the way, did you do something like `git push --force`? If you are still interested in this PR, I can restructure it for merging.

**Comment by fulls1z3** [comment]
Hi @zxj20, I had to perform some rebase operations, and this is why we have merge issues. About the PR, i'm interested. I was busy on the last weeks with server-side rendering for `@nglibs/meta` but I made the release just a few hours ago. So, this PR is in the priority list now.

**Comment by zjx20** [comment]
@fulls1z3 I had restructured the commits, but I can't figure out why *appveyor* failed. Tests were actually passed on *appveyor* and work well in my local env.

**Comment by fulls1z3** [comment]
Hi @zjx20, I know why it fails: types for lodash has problems on versions higher than `^4.14.55`.

The solution depends on `^4.14.55` which causes the build problem. When the version is fixed to `4.14.55` (instead of `^4.14.55`) the problem will disappear. I'll manage this.

**Comment by zjx20** [comment]
Got it, thank you for telling me that.

**Comment by fulls1z3** [comment]
You're welcome, and thanks for your awesome contributions. I'll merge it within the day.

**Comment by fulls1z3** [comment]
Hi @zjx20, I've successfully merged and refactored your PR. Now it works like a charm, thanks for all your efforts.

Now, we have such an use case, and I would like to listen to your ideas:

- Let's say we have two config files: `config.local.json`, and `http://some-api-provider.com/api/get-settings`.
- We pass these locations to the `paths` property of `ConfigMergedHttpLoader`.
- We want `ConfigService` processes those two config files sequentally, and replace the API URL `some-api-provider` with a value from `config.local.json`.

In other words:

#### config.local.json
```json
{
  "api-endpoint": "http://some-api-provider/api"
}
```

#### app.module
```TypeScript
const configFactory(http: Http): ConfigLoader {
  return new ConfigMergedHttpLoader(http, ['./config.local.json', '{{api-endpoint}}/get-settings'])
}
```

What would be your suggestion?

**Comment by fulls1z3** [comment]
Hi @unlight,

About `lodash`:  I found out a way to exclude it from the core of `config` project.

On the other hand, I've just checked the functionality on your fork and it looks great - especially the Pipe. Shall we add them to `nglibs/config` too?

**Comment by unlight** [comment]
Hi, I'm still working and testing, my concern is about `getSettings` and `get` methods.
They do same job.

`getSettings`:
\+ throws exception
\+ can return entire config
\- restricted 2 level deep
\- no default value
\- long name

`get`:
\+ no restrictions in deep
\+ default value
\+ shorter name
\- no exceptions
\- no entire config

**Comment by fulls1z3** [comment]
Hi @unlight, I have separated all loaders except `ConfigStaticLoader` from the project, so it no longer depends on `lodash` and on `@angular/http`. It will keep bundles sizes at minimal level.

Anyone who needs a certain loader can import from a variety of loaders available there:
- `Http`
- `Fs` (for server-side)
- `Parallel` (a.k.a `Merged`), `Series` (a.k.a `Merged`)
- `Universal` (switches between `Http`/`Fs` depending on server/browser platform, with server/browser caching)

If we talk about your concerns, `getSettings` and `get` do the same work. I'll work on it and I think it's possible to have all features in one method. Stay tuned, I'll update the repo shortly.

Meanwhile, I'll be glad if you can contribute with your ideas/feedback.

**Comment by fulls1z3** [comment]
@unlight you can check now the latest commits and provide feedback.

**Comment by unlight** [comment]
@fulls1z3 
Nice job :+1: 

Looks like you implemented all, I do not have any other ideas or requirements.
I'll take a closer look in next few days.

**Comment by unlight** [comment]
One thing.
It is impossible to get falsy values (zero, null) due to this [relaxed check](https://github.com/ngx-config/core/blob/c479a9d3931358fd7f0496f2053f55101f18e91d/src/config.service.ts#L28-L29)

I've created PR to fix this, take a look: #14
